src/ssl.c

```diff
@@ -12036,6 +12036,172 @@ static int wolfSSL_remove_ciphers(char* list, int sz, const char* toRemove)
+            enabledcs = GetCipherNameInternal(cipherSuite0, cipherSuite);
+        } 
+        else if (!tls13Only && cipherSuite0 != TLS13_BYTE) {
+            enabledcs = GetCipherNameInternal(cipherSuite0, cipherSuite);
+        } 
+        else
+            continue;
+        
+        if (XSTRNCMP(enabledcs, "None", XSTRLEN(enabledcs)) != 0) {
+            len += (word32)XSTRLEN(enabledcs) + 2;
+        }
+    }
+    
+    len += listsz + 2;
+    
+    /* build string */
+    if (len > (listsz + 2)) {
+        locallist = (char*)XMALLOC(len, ctx->heap,
+                                           DYNAMIC_TYPE_TMP_BUFFER);
+        /* sanity check */
+        if (!locallist)
+            return NULL;
+        
+        XMEMSET(locallist, 0, len);
+        
+        head = locallist;
+        
+        if (!tls13Only)
+        {
+            /* always tls13 suites in the head position */
+            XSTRNCPY(locallist, list, len);
+            locallist += listsz;
+            *locallist++ = ':';
+            *locallist = 0;
+            len -= listsz + 1;
+        }
+        
+        for(idx = 0; idx < suites->suiteSz; idx++) {
+            cipherSuite0 = suites->suites[idx];
+            cipherSuite  = suites->suites[++idx];
+            
+            if (tls13Only && cipherSuite0 == TLS13_BYTE) {
+                enabledcs = GetCipherNameInternal(cipherSuite0, cipherSuite);
+            }
+            else if (!tls13Only && cipherSuite0 != TLS13_BYTE) {
+                enabledcs = GetCipherNameInternal(cipherSuite0, cipherSuite);
+            } 
+            else
+                continue;
+            
+            ianasz = (int)XSTRLEN(enabledcs);
+            if (ianasz + 1 < len) {
+                XSTRNCPY(locallist, enabledcs, len);
+                locallist += ianasz;
+                
+                *locallist++ = ':';
+                *locallist = 0;
+                len -= ianasz + 1;
+            }
+            else{
+                XFREE(locallist, ctx->heap, DYNAMIC_TYPE_TMP_BUFFER);
+                return NULL;
+            }
+        }
+        
+        if (tls13Only) {
+            XSTRNCPY(locallist, list, len);
+            locallist += listsz;
+            *locallist = 0;
+        }
+        
+        return head;
+    } 
+    else 
+        return NULL;
+}
+
+/*
+ * check if the list has TLS13 and pre-TLS13 suites
+ * @param list cipher suite list that user want to set
+ * @return mixed: 0, only pre-TLS13: 1, only TLS13: 2
+ */
+static int CheckcipherList(const char* list)
+{
+    int ret;
+    int findTLSv13Suites = 0;
+    int findbeforeSuites = 0;
+    byte cipherSuite0;
+    byte cipherSuite1;
+    int flags;
+    char* next = (char*)list;
+    
+    do {
+        char*  current = next;
+        char   name[MAX_SUITE_NAME + 1];
+        word32 length;
+        
+        next   = XSTRSTR(next, ":");
+        length = min(sizeof(name), !next ? (word32)XSTRLEN(current) /* last */
```

Comment:
Make it the min(sizeof(name) - 1, ...) and there will always be space for the '\0'.

Reply:
Oh, Thanks.

Reply:
Change code to the following causes error "‘__builtin___strncpy_chk’ specified bound depends on the length of the source argument" on gcc-9.3 that I am using as default gcc compiler,

        length = min(sizeof(name) - 1, !next ? (word32)XSTRLEN(current) /* last */
                                         : (word32)(next - current));
        XSTRNCPY(name, current, length);
        name[length] = 0;

 gcc-7 and gcc-8 are ok. Any idea?

Reply:
Not using the ternary operator appears to fix it.
1. Calculate the current length
2. assign sizeof(name) - 1 to length and
3. assign current length to length when current length is smaller

Reply:
Thanks!